### PR TITLE
Adopt the CLI UI kit in the `shopify theme open` command

### PR DESCRIPTION
**Pull request by karreiro**
### WHY are these changes introduced?

Fixes:
- https://github.com/Shopify/internal-cli-foundations/issues/496
- https://github.com/Shopify/internal-cli-foundations/issues/497

### WHAT is this pull request doing?

This PR has two main parts, the REST Admin API client and the `shopify theme open` changes. Let's approach these changes by taking a look at modified files:

- **`.vscode/launch.json`** - it introduces a handy debugger for themes

- **`packages/cli-kit/src/api/admin*`** - here we're introducing the thin client to handle REST API calls, at the [@shopify/cli-kit](https://github.com/Shopify/cli/blob/main/docs/cli/architecture.md#modules-packages) level

- **`packages/theme/src/cli/utilities/themes-api*`** - here we're introducing themes API-specific logic

- **`packages/theme/src/cli/commands/theme/open*`**
  **`packages/theme/src/cli/services/open*`** - the `shopify theme open` command was updated following the [convention](https://github.com/Shopify/cli/blob/main/docs/cli/conventions.md#naming-and-directory-conventions-1) of having a service module with the same name

- **`packages/theme/src/cli/models/theme.ts`** - it's a thin data-structure that represents theme

- **`packages/theme/src/cli/utilities/theme-selector*`**
- **`packages/theme/src/cli/utilities/theme-urls*`** - here we're introducing utility functions to support our `services`

### How to test your changes?

- Run `pnpm run shopify theme open -t <theme ID>`
- Run `pnpm run shopify theme open -t <theme name>`
- Run `pnpm run shopify theme open -l`
- Run `pnpm run shopify theme open -d`
- Run `pnpm run shopify theme open`

**Notice**: The [prototype](https://user-images.githubusercontent.com/1079279/210743403-7a41ab62-56ef-43f9-af8c-14444f052857.png) doesn't have a header, however the CLI UI kit API requires it. I'm checking with DX, so feel free to ignore that detail for now :)

**Before**
Screenshot 2023-01-05 at 14 33 09

**After**
Screenshot 2023-01-05 at 14 34 09

### Post-release steps

None.

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [x] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've considered possible [documentation](https://shopify.dev) changes
- [x] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

**Comment by karreiro** [comment]
👋 Hey folks,

Feel free to the [benchmark report](https://github.com/Shopify/cli/pull/1014#issuecomment-1372236353) - I've noticed it's happening on other PRs as well (I don't think it's related to these changes, still I'm double-checking)

**Review by Poitrin — CHANGES_REQUESTED** [review]
Commands mentioned in 🎩 instructions work as expected.

Is `pnpm run shopify theme open -e` also supposed to work already?
Without the `-t <theme>` flag, I get …
```
╭─ error ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│                                                                                                                                                                  │
│  Cannot read properties of undefined (reading 'role')                                                                                                            │
│                                                                                                                                                                  │
│  To investigate the issue, examine this stack trace:                                                                                                             │
│    at themePreviewUrl (theme/src/cli/utilities/theme-urls.ts:8)                                                                                                  │
│      if (theme.role === 'live') {                                                                                                                                │
│    at open (theme/src/cli/services/open.ts:17)                                                                                                                   │
│      const previewUrl = themePreviewUrl(theme, adminSession)                                                                                                     │
│    at processTicksAndRejections (node:internal/process/task_queues:95)                                                                                           │
│    at run (theme/src/cli/commands/theme/open.ts:42)                                                                                                              │
│      await open(adminSession, flags)                                                                                                                             │
│                                                                                                                                                                  │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
```

File: packages/theme/src/cli/utilities/themes-api/headers.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+function asInt(strValue: string) {
+  const intValue = parseInt(strValue, 10)
+
+  if (isNaN(intValue)) {
+    return null
+  }
+
+  return intValue
+}
```
Comment (isaacroldan) [inline packages/theme/src/cli/utilities/themes-api/headers.ts:48]: we already have a function called `tryParseInt` on cli-kit, can we reuse that one instead?

File: .changeset/polite-schools-burn.md [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+'@shopify/cli-kit': patch
+'@shopify/theme': patch
+---
+
+- Introduce REST Admin API client on `@shopify/cli-kit`
+- Improve the DX of the `shopify theme open` command, by adopting the CLI UI kit
```
Comment (isaacroldan) [inline .changeset/polite-schools-burn.md:7]: All these changes feel more like a "minor" than a patch, what do you think?

Comment (karreiro) [reply .changeset/polite-schools-burn.md:7]: @isaacroldan This PR have many additions to provide a more polished DX. Still, this PR is also not introducing any feature or functional change in the commands, so, from the users perspective, I'd still consider it a "minor" change.

**Review by Poitrin — APPROVED** [review]
Besides the `packages/theme/oclif.manifest.json` merge conflict that sneaked in, all 🎩 steps passed now, and code LGTM – great work :+1:

**Comment by karreiro** [comment]
I just rebased the branch to solve conflicts.

File: packages/theme/src/cli/utilities/theme-selector/filter.test.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    const filtered = filterThemes(store, themes, filter)
+
+    // Then
+    expect(filtered).toHaveLength(1)
+    expect(filtered[0]!.name).toBe('theme 3')
+  })
+
+  test('filters the development theme', async () => {
+    /**
+     * TODO: Return _your_ development theme.
```
[the remark this answers is not in the document]
Comment (karreiro) [reply packages/theme/src/cli/utilities/theme-selector/filter.test.ts:35]: https://github.com/Shopify/internal-cli-foundations/issues/509

File: packages/cli-kit/src/private/node/ui/components/Subdued.tsx [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+interface Props {
+  subdued: string
+}
+
+/**
+ * `Subdued` displays some text that represents a sub-process
+ */
+const Subdued: React.FC<Props> = ({subdued}: React.PropsWithChildren<Props>): JSX.Element => {
+  return <Text>{chalk.dim(subdued)}</Text>
```
Comment (matteodepalo) [inline packages/cli-kit/src/private/node/ui/components/Subdued.tsx:13]: You can use the `Text` component property `dimColor` here so `<Text dimColor>{subdued}</Text>`

File: packages/cli-kit/src/private/node/ui/components/Subdued.tsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import {Text} from 'ink'
+import chalk from 'chalk'
+import React from 'react'
+
+interface Props {
+  subdued: string
+}
+
+/**
+ * `Subdued` displays some text that represents a sub-process
```
Comment (matteodepalo) [inline packages/cli-kit/src/private/node/ui/components/Subdued.tsx:10]: Is a `sub-process` semantically linked to subdued text? What if instead we said that `Subdued` displays some text with subdued colors?

File: packages/cli-kit/src/private/node/ui/components/Subdued.tsx (hunk 4 above)
Comment (karreiro) [reply packages/cli-kit/src/private/node/ui/components/Subdued.tsx]: +💯 @matteodepalo, thank you! I've applied the enhancement in a new commit :)

**Review by mgmanzella — APPROVED** [review]

**Comment by karreiro** [comment]
(I've rebased the branch following the refactor in the Admin public API)

**Comment by karreiro** [comment]
(I've refreshed `oclif.manifest.json` and I've rebased the branch following the refactor in the Admin public API)

**Pull request merged**